xarray/core/dataset.py

```diff
@@ -793,20 +805,28 @@ def copy(self, deep=False, data=None):
         if data is None:
             variables = OrderedDict((k, v.copy(deep=deep))
                                     for k, v in iteritems(self._variables))
-            # skip __init__ to avoid costly validation
-            return self._construct_direct(variables, self._coord_names.copy(),
-                                          self._dims.copy(), self._attrs_copy(),
-                                          encoding=self.encoding)
-        if not utils.is_dict_like(data):
+        elif not utils.is_dict_like(data):
             raise ValueError('Data must be dict-like')
-        elif not set(data) <= set(self._variables.keys()):
-            raise ValueError('All keys in `data` must be'
-                             'contained in the original dataset')
-
-        variables = OrderedDict((k, self._variables[k].copy(data=v))
-                                for k, v in iteritems(data))
-        return self._subset_with_all_valid_coords(variables, set(),
-                                                  attrs=self.attrs.copy())    
+        else:
+            var_keys = set(self.data_vars.keys())
+            data_keys = set(data.keys())
+            keys_not_in_vars = data_keys - var_keys
+            if keys_not_in_vars:
+                raise ValueError('Data must only contain variables in original '
+                                 'dataset. Extra variables: {}'
+                                 .format(keys_not_in_vars))
+            keys_missing_from_data = var_keys - data_keys
+            if keys_missing_from_data:
```

Comment:
you could potentially consolidate these checks into `if `data_keys != var_keys`, but this is probably a more informative error message :)

Reply:
yep I wrote them that way on purpose for the error message.